### PR TITLE
Large scale suites & CPP traces

### PR DESCRIPTION
**Pull request by thomasuster**
When projects are large enough they test runner will crash. I fixed this by dereferencing already executed test suits.

When null exceptions occur in C++ the test runner fails to catch the exception and the stack trace is lost. I conditionally removed the redirection of haxe.trace to allow for trace debugging.

### CODE REVIEW

**Comment by thomasuster** [comment]
bump

**Comment by thomasuster** [comment]
bump

File: src/massive/munit/TestRunner.hx [hunk 1]
```diff
@@ -100,6 +100,8 @@ class TestRunner implements IAsyncDelegateObserver
     public var completionHandler:Bool -> Void;
 
     public var clientCount(get_clientCount, null):Int;
+    public var testSuites:Array<TestSuite>;
```
Comment (elsassph) [inline src/massive/munit/TestRunner.hx]: Does it need to be public?

**Comment by mikestead** [comment]
Not entirely happy with removing the capture trace for cpp target. This breaks the output of a TestResultClient which existing setups may depend upon. I can understand why you did it, just has to be backwards compatible when you make changes like this.

Also I've not seen the issue you're describing with test runner crashing, do you know why dereferencing fixes the issue? I assume some memory / GC problem? Don't suppose there's an open set of tests where we can see this happening so I can verify these changes do the trick?

**Comment by mikestead** [comment]
In the short term you could cleanup any memory allocated in your tests using `@After` or `@AfterClass` functions.

File: src/massive/munit/TestRunner.hx (hunk 1 above)
Comment (thomasuster) [inline src/massive/munit/TestRunner.hx]: I exposed it for my change in TestRunnerTest.hx, I could instead add a new method for the check but this seemed simplest.

**Comment by thomasuster** [comment]
@mikestead 

Trace change: Yea, if someone was doing something fancy with the test output this change would break their code. I do think the problem this fixes is quite a bummer for most devs, it took me a long time to track down what was going on. My preference would be to make people who are doing something fancy with the output to add a command line argument to enable the legacy behavior (so the novices won't have to). What do you think?

Memory change: Dereferencing fixes the issue because it allows the hxcpp mark & sweep algorithm to sweep away the dereferences objects. Because testSuites is a member variable the hxcpp crawler keeps marking the already built and run tests.

**Comment by mikestead** [comment]
A breaking change no matter what it is will require a major version bump. That's not something that we'll want for this so it will need to be backwards compatible, at least for now.

RE memory allocation, my thinking is that you're not cleaning up memory allocated in your tests, so for now you can do that to fix the issue? That might explain why no one else has run into this problem? We have some reasonably large test suites and not seen this. That's not to say your fix here isn't also valid, and most likely preferred.

Anyway I'll give these changes a test today, I think the trace one may need to be dropped from this pull request or wrapped in a different flag (perhaps be useful for other platforms too?). I'll see if I can come up with a better suggestion after some testing.

**Comment by thomasuster** [comment]
@mikestead Good feedback Mike. Let me resubmit focusing just on the GC change. Please hold off testing until then.

**Comment by mikestead** [comment]
Sure, you don't need to close though, just update this one. FYI I don't think exposing the property as public is worth the test in this case. I'd just null the array value like you did without the test. Not ideal but would prefer to keep some things internal.

**Comment by thomasuster** [comment]
https://github.com/massiveinteractive/MassiveUnit/pull/99

**Pull request closed without merging**